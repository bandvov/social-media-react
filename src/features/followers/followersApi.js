import { api } from "../api";

export const fetchFollowers = async ({ userId, ...params }) => {
  return await api.get(`/users/${userId}/followers`, {
    params,
  });
};

export async function removeFollower(followeeId) {
  return await api.delete(`/followers/${followeeId}`);
}
