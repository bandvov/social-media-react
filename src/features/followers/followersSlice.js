import { createSlice } from "@reduxjs/toolkit";

const followersSlice = createSlice({
  name: "followers",
  initialState: {
    data: [
      {
        id: 12,
        username: "test",
        email: "bla@bla.bla",
        profile_pic:
          "https://upload.wikimedia.org/wikipedia/commons/thumb/c/cb/Elon_Musk_Royal_Society_crop.jpg/440px-Elon_Musk_Royal_Society_crop.jpg",
        bio: "test test test bio",
      },
      {
        id: 26,
        username: "test",
        email: "bla@bla.bla",
        profile_pic:
          "https://upload.wikimedia.org/wikipedia/commons/thumb/0/06/Giselle_of_Aespa%2C_July_25%2C_2024_%282%29.png/526px-Giselle_of_Aespa%2C_July_25%2C_2024_%282%29.png",
      },
      {
        id: 883,
        username: "test",
        email: "bla@bla.bla",
      },
    ],
    page: 1,
    isLoading: false,
    hasMore: true,
    error: null,
  },
  reducers: {
    fetchFollowersRequest: (state) => {
      state.isLoading = true;
      state.error = "";
    },
    fetchFollowersSuccess: (state, action) => {
      state.isLoading = false;
      state.data = [...state.data, ...action.payload.followers];
      state.page = action.payload.nextPage;
      state.hasMore = action.payload.hasMore;
    },
    fetchFollowersFailure: (state, action) => {
      state.isLoading = false;
      state.error = action.payload;
    },
    removeFollowerRequest: (state) => {
      state.loading = true;
      state.error = false;
    },
    removeFollowerSuccess: (state, action) => {
      state.loading = false;
      state.followers = [
        ...state.followers.filter((f) => {
          return f.id !== action.payload;
        }),
      ];
    },
    removeFollowerFailure: (state, action) => {
      state.loading = false;
      state.error = action.payload;
    },
  },
});

export const {
  fetchFollowersRequest,
  fetchFollowersSuccess,
  fetchFollowersFailure,
  removeFollowerFailure,
  removeFollowerRequest,
  removeFollowerSuccess,
} = followersSlice.actions;

export default followersSlice.reducer;
