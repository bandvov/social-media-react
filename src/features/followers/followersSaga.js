import { call, put, takeLatest, select } from "redux-saga/effects";
import {
  fetchFollowersRequest,
  fetchFollowersSuccess,
  fetchFollowersFailure,
  removeFollowerFailure,
  removeFollowerRequest,
  removeFollowerSuccess,
} from "./followersSlice";
import { fetchFollowers, removeFollower } from "./followersApi";

function* handleFetchFollowers(action) {
  try {
    const { page } = yield select((state) => state.followers);
    const data = yield call(fetchFollowers, { page, ...action.payload });
    yield put(
      fetchFollowersSuccess({
        followers: data.followers,
        nextPage: page + 1,
        hasMore: data.hasMore,
      })
    );
  } catch (error) {
    yield put(fetchFollowersFailure(error.message));
  }
}
function* handleRemoveFollower(action) {
  try {
    yield call(removeFollower, action.payload);
    yield put(removeFollowerSuccess(action.payload));
  } catch (error) {
    yield put(removeFollowerFailure(error.message));
  }
}

export default function* followersSaga() {
  yield takeLatest(fetchFollowersRequest.type, handleFetchFollowers);
  yield takeLatest(removeFollowerRequest.type, handleRemoveFollower);
}
