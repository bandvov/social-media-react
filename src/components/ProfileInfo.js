import {
  Avatar,
  Box,
  Container,
  IconButton,
  Link,
  Menu,
  MenuItem,
  Typography,
} from "@mui/material";
import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { fetchUserProfileRequest } from "../features/user/userSlice";
import { Link as RouterLink } from "react-router-dom";
import MoreVertIcon from "@mui/icons-material/MoreVert";
import { useTranslation } from "react-i18next";
import { removeFollowerRequest } from "../features/followers/followersSlice";

export default function ProfileInfo({ userId }) {
  const { t } = useTranslation();

  const dispatch = useDispatch();
  const { profile } = useSelector((state) => state.user);
  const {
    email,
    posts_count,
    profile_pic,
    followees_count,
    followers_count,
    created_at,
    bio,
  } = profile;
  const [menuAnchor, setMenuAnchor] = React.useState(null);

  useEffect(() => {
    if (userId) {
      dispatch(fetchUserProfileRequest(userId));
    }
  }, [dispatch, userId]);

  const handleMenuOpen = (event) => {
    setMenuAnchor(event.currentTarget);
  };

  const handleMenuClose = () => {
    setMenuAnchor(null);
  };

  const handleRemoveFollower = () => {
    handleMenuClose();
    dispatch(removeFollowerRequest(userId));
  };

  return (
    <Container maxWidth="md">
      <Box borderBottom={1} borderColor="divider" mb={2}>
        <Box
          display="flex"
          alignItems="center"
          justifyContent="space-between"
          pb={2}
        >
          {/* Avatar and Name */}
          <Box display="flex" alignItems="center">
            <Avatar
              src={profile_pic}
              alt={email}
              sx={{ width: 64, height: 64, mr: 2 }}
            />
            <Box>
              <Typography color="text.primary" variant="h6">
                {email}
              </Typography>
              <Typography color="text.primary" variant="body2">
                Joined: {new Date(created_at).toLocaleDateString()}
              </Typography>
            </Box>
          </Box>

          {/* Stats */}
          <Box display="flex" alignItems="center" gap={4}>
            <Typography color="text.primary" variant="body2">
              <Link component={RouterLink} to={`/user/${profile?.id}/posts`}>
                <strong>{posts_count}</strong> Posts
              </Link>
            </Typography>
            <Typography color="text.primary" variant="body2">
              <Link
                component={RouterLink}
                to={`/user/${profile?.id}/followers`}
              >
                <strong>{followers_count}</strong> Followers
              </Link>
            </Typography>
            <Typography color="text.primary" variant="body2">
              <Link
                component={RouterLink}
                to={`/user/${profile?.id}/followees`}
              >
                <strong>{followees_count}</strong> Following
              </Link>
            </Typography>
          </Box>

          {/* Action Menu */}
          <IconButton onClick={handleMenuOpen}>
            <MoreVertIcon />
          </IconButton>
          <Menu
            anchorEl={menuAnchor}
            open={Boolean(menuAnchor)}
            onClose={handleMenuClose}
          >
            {userId && (
              <MenuItem onClick={handleMenuClose}>
                <Link component={RouterLink} to="/login">
                  Edit profile
                </Link>
              </MenuItem>
            )}
            <MenuItem onClick={handleMenuClose}>{t("follow")}</MenuItem>
            <MenuItem onClick={handleRemoveFollower}>{t("unfollow")}</MenuItem>
          </Menu>
        </Box>
        {bio && (
          <Box pb={2}>
            <Typography color="text.primary">Bio: {bio}</Typography>
          </Box>
        )}
      </Box>
    </Container>
  );
}
