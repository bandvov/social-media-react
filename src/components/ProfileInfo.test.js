import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import ProfileInfo from "./ProfileInfo";
import {
  fetchUserProfileRequest,
  setInitialUserState,
} from "../features/user/userSlice";
import { removeFollowerRequest } from "../features/followers/followersSlice";
import "@testing-library/jest-dom";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    t: (key) => key, // Mock translation function
  }),
}));

jest.mock("react-router-dom", () => ({
  Link: ({ children }) => <div>{children}</div>,
  useParams: () => ({ userId: "1" }),
  useNavigate: jest.fn(),
}));
// Mock Redux
jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

// Mock Dispatch Function
const mockDispatch = jest.fn();

describe("ProfileInfo Component", () => {
  let store;
  const mockStore = configureStore([]);

  beforeEach(() => {
    store = mockStore({
      user: {
        profile: {
          id: 1,
          email: "test@example.com",
          posts_count: 5,
          profile_pic: "https://via.placeholder.com/64",
          followees_count: 10,
          followers_count: 8,
          created_at: "2024-01-01T00:00:00Z",
          bio: "Test Bio",
          is_follower: true,
          is_followee: false,
        },
      },
      auth: { user: { id: 2 } },
    });

    jest
      .spyOn(require("react-redux"), "useDispatch")
      .mockReturnValue(mockDispatch);
    jest
      .spyOn(require("react-redux"), "useSelector")
      .mockImplementation((callback) => {
        return callback(store.getState());
      });
  });

  test("renders user profile data correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <ProfileInfo userId={1} />
      </Provider>
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByText("test@example.com")).toBeInTheDocument();
    expect(screen.getByText(/Joined:/)).toBeInTheDocument();
    expect(screen.getByText("5")).toBeInTheDocument();
    expect(screen.getByText("8")).toBeInTheDocument();
    expect(screen.getByText("10")).toBeInTheDocument();
    expect(screen.getByText("Bio: Test Bio")).toBeInTheDocument();
  });

  test("dispatches actions on mount", () => {
    render(
      <Provider store={store}>
        <ProfileInfo userId={1} />
      </Provider>
    );

    expect(mockDispatch).toHaveBeenCalledWith(setInitialUserState());
    expect(mockDispatch).toHaveBeenCalledWith(fetchUserProfileRequest(1));
  });

  test("opens and closes menu", () => {
    render(
      <Provider store={store}>
        <ProfileInfo userId={1} />
      </Provider>
    );

    const menuButton = screen.getByRole("button");
    fireEvent.click(menuButton);

    expect(screen.getByText("unfollow")).toBeInTheDocument();

    fireEvent.click(screen.getByText("unfollow"));

    expect(mockDispatch).toHaveBeenCalledWith(removeFollowerRequest(1));
  });
 
});
