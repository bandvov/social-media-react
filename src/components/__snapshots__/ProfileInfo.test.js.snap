// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileInfo Component renders user profile data correctly 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthMd css-u4uvlt-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-1fkybvm"
    >
      <div
        class="MuiBox-root css-cmyu5s"
      >
        <div
          class="MuiBox-root css-70qvj9"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-yjdvsz-MuiAvatar-root"
          >
            <img
              alt="test@example.com"
              class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
              src="https://via.placeholder.com/64"
            />
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-vvtk0a-MuiTypography-root"
            >
              test@example.com
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body2 css-1yrr6qh-MuiTypography-root"
            >
              Joined: 
              01.01.2024
            </p>
          </div>
        </div>
        <div
          class="MuiBox-root css-thflsa"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-1yrr6qh-MuiTypography-root"
          >
            <div>
              <strong>
                5
              </strong>
               Posts
            </div>
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-1yrr6qh-MuiTypography-root"
          >
            <div>
              <strong>
                8
              </strong>
               Followers
            </div>
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-1yrr6qh-MuiTypography-root"
          >
            <div>
              <strong>
                10
              </strong>
               Following
            </div>
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="MoreVertIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2m0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2m0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2"
            />
          </svg>
        </button>
      </div>
      <div
        class="MuiBox-root css-1ebnygn"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-4leci6-MuiTypography-root"
        >
          Bio: 
          Test Bio
        </p>
      </div>
    </div>
  </div>
</div>
`;
